R/set-record-option.R

```diff
@@ -0,0 +1,110 @@
+#'
+#' Note that only estimation, data, and table records are currently supported.
+#'
+#' Notes on `value`:
+#'  - We infer whether to use `option_value` or `option_flag` for *new* options based on the *class* of `value`:
+#'     - If you are trying to add an option with an assigned value (i.e. `MAXEVAL = 10`), `value` must be a character string or numeric value
+#'     - If you are trying to add a flag option (i.e. `NOABORT`), `value` should be logical (`T`/`F`).
+#'  - If `value` is set to `NULL`, the option will be removed (if found). `FALSE` is also acceptable for removing an existing option of class `nmrec_option_flag`.
+#'
+#' @param record An [nmrec_record] object.
+#' @param name Name of record to select. Any valid spelling of the option name
+#'   is allowed.
+#' @param value value to assign to option.
+#'
+#' @seealso `nmrec::option` selecting records of a given type.
+#'
+#'
+set_record_option <- function(record,
+                              name,
+                              value) {
+  stopifnot(inherits(record, "nmrec_record"))
+  stopifnot(length(name) == 1, is.character(name), nzchar(name))
+  stopifnot(record[["name"]] %in% c("estimation", "data", "table"))
+
+
+  name_map <- get(paste0(record[["name"]], "_option_names"))
+  name_resolved <- get0(tolower(name), name_map)
+  rec_opt_prev <- get_record_option(record, name)
+
+
+  if (is.null(name_resolved) && is.null(rec_opt_prev)) {
+    abort(
+      sprintf("%s is not a valid option for %s blocks", name, record[["name"]]),
+      nmrec_error()
+    )
+  }
+
+  # Create new record option
+  if (is.null(rec_opt_prev)) {
+    if (is.null(value)) {
+      abort(
+        sprintf("Option %s does not exist in this record. `NULL` is only valid for removing existing options.", name),
+        nmrec_error()
+      )
+    } else {
+      opt_type <- if (is.logical(value)) "flag" else "value"
+
+      if (opt_type == "value") {
+        stopifnot(is.numeric(value) || is.character(value))
+        opt <- option_value$new(name_resolved, name, value = value)
+      } else if (opt_type == "flag") {
+        stopifnot(is.logical(value))
+        opt <- option_flag$new(name_resolved, name)
+      }
+    }
+
+    new_opt_lst <- list(elem_whitespace(" "), opt)
+
+    record$parse()
+    record$values <- append(record$values, new_opt_lst, after = find_record_end(record) - 1)
+  } else {
+    # Modify existing option
+    if (inherits(rec_opt_prev, "nmrec_option_value")) { # If option_value
+      stopifnot(is.numeric(value) || is.character(value) || is.null(value))
+      rec_opt_prev$value <- value
+    } else if (inherits(rec_opt_prev, "nmrec_option_flag")) { # If option_flag
+      stopifnot(is.logical(value) || is.null(value))
+      if (isFALSE(value)) value <- NULL # Accept NULL or FALSE for removing option
+      rec_opt_prev$value <- value
+    } else if (inherits(rec_opt_prev, "nmrec_option_pos")) { # If option_pos
```

Comment:
Please drop this handling and just support the flag and value options.  If limited to existing options, positional options can already be conveniently grabbed and modified.

Reply:
This would prevent `set_record_option(record, name = "list1", value = new_vec)` from working. `get_record_option(record, name = "list1")` works, so I feel like we should support this as well with `set_record_option`. It does seem odd that we would support modifying positional options and not creating new ones though, so I am a little conflicted

Reply:
As I've said, my vote is very much for this helper to be limited to flag/value options.  The case you're covering can just as conveniently be handled by grabbing the option and doing `opt$value <- val`.  We don't need another way to do the same thing.

(My goal here is to keep these helpers simple and straightforward while handling the common case of flag/value options.)